### ISSUE TITLE
AI behaves wrong in scenarios

### ISSUE THREAD

**Issue by hessenfarmer** [widelands/widelands#4558]
Some recent playtesting showed the AI behaving differently in Scenarios than in normal single player.
Especially it does not expand / does not build military buildings. This is going away with a save and reload.

**Comment by Noordfrees** [comment]
The bug is probably related to this line:
https://github.com/widelands/widelands/blob/master/src/logic/game.cc#L248
A player's AI defaults to "" unless every scenario designer remembers to set one explicitly, and this is not a valid AI name and results in use of an arbitrary AI implementation. This code is only run when starting a new scenario game.

**Comment by hessenfarmer** [comment]
I experimented a bit with the AI setting an AI didn't solve the issue. Only saving reloading does the trick. I think it might be the case for normal singleplayer as well but is not visible there due to saving reloading for the replay.

**Comment by Noordfrees** [comment]
So if we found a way to make replays work for scenarios (re #4531) that might fix this issue as well…

**Comment by hessenfarmer** [comment]
not sure need to investigate the root cause first

**Comment by hessenfarmer** [comment]
I am still struggling to identify what is the difference in the sequence to initilaise a singleplayer scenario, a normal singleplayer game and a savegame. I believe the problem lies in this sequence. As the AI is working after a save and reload. (this is done in a normal game to initialize the replay as far as I know so the problem might exist there as well but gets masked. maybe it's worth trying this with trainingwheels active as there are no replays with training wheels as well.

**Comment by hessenfarmer** [comment]
I am not 100% sure anymore this is a general behaviour as with p1.see_all activated I watched some scenario AI and they were expanding pretty well and should not be different from normal game which I will try to confirm. 
But in fri03 they are only expanding after save reload. In the enabled log after reload the Buildings were shortly listed as being forced which is weird as I don't understand how this could happen. However it may be related to the fact that there are very few building spots available at the beginning and so the AI can't decide what to build. Will investigate this as well.

**Comment by hessenfarmer** [comment]
I did further testing:
- It is a general behaviour
- Fri 03 is more affected as the Ai has very little buildspace in the beginning
- Even with enough build space the AI is only building small Military Buildings (if at all) until Save/reload
- The AI is identified as Normal if not set explicitly. (Can't tell if it is initialized properly) 
- in normal games (at least network games) the Ai behaves as it should, and expands 
- using the last of the AI branches (#4618) did not help
- I was not able to test b21 or b20 as the fri03 map isn't compatible 

I used the p1.see_all = true line to view the AI in scenario mode and a local network setup with AI only to observe it in non scenario mode

@Noordfrees for me this is a major bug, (although it might be an old one) as all scenarios are non deterministic with this and especially fri03 is highly affected.
I strongly believe it is related to the initialization sequence of a game. Especially the point in time when the computerplayer is initialized. However I am struggling to identify the proper sequences (and differences) in the given code as I am not that familiar with all Constructs of C++. So any help is appreciated

**Comment by Noordfrees** [comment]
We have like 4 or 5 different ways of initializing a game so this is not trivial. I have tried just enabling the replay writer for scenarios so a savegame is initially created and loaded like for normal games but this didn't help. I agree that this bug could be a showstopper. I do suspect though that some (replay-related) save-reloading magic affects the AI behaviour, Will try to investigate this further tomorrow

**Comment by hessenfarmer** [comment]
> 
> 
> We have like 4 or 5 different ways of initializing a game so this is not trivial. I have tried just enabling the replay writer for scenarios so a savegame is initially created and loaded like for normal games but this didn't help. I agree that this bug could be a showstopper. I do suspect though that some (replay-related) save-reloading magic affects the AI behaviour, Will try to investigate this further tomorrow

Many thanks. Please let me know if I can be of any help.
regarding the replays I am not sure whether we are still doing the save reload. I have distant memories this got changed for b20 or so as to reduce loading times for big games.

**Comment by Noordfrees** [comment]
I believe there was a time when the save-reloading happened twice and this was changed. Now it still does it but only once.

---

The game creation pathways, just for reference:
- New MP game: `Game::init_newgame` then `Game::run`
- Load MP game: `Game::init_savegame` then `Game::run`
- New SP game: `Game::init_newgame` then `Game::run`
- Load SP game: `Game::run_load_game` (which calls `Game::run`)
- SP Scenario (from tutorial/scenario select, when selecting a scenario in SPG mode, and with `--scenario`): `Game::run_splayer_scenario_direct` (which calls `Game::run`)
- Replay: plain `Game::run`

**Comment by hessenfarmer** [comment]
Update:
I added a saveloading cycle as else branch to the writereplay initialization (https://github.com/widelands/widelands/blob/a846817ed8a75d18ad4a84a8fc28c5375efc15a4/src/logic/game.cc#L638) and can confirm it doesn't help in this stage of the game. 

However I copmpared the output in my stdout.txt from different attempts to start a game.     
   
This led to one difference I could identify in the output: It seems that in scenarios the player isn't initialized properly or completely as there is no starting building until the scenario scripts start to run and places them. Maybe the culprit lies in there.

**Comment by Noordfrees** [comment]
This should not matter – in both cases the script that places the initial buildings for all players is run at the gametime 1ms. 
It seems like sometimes players are deleted and re-created during saveloading and sometimes not (even when adding the additional saveload cycle), and I suspect this might have something to do with the AI not being initialized correctly. I still did not find out much useful though…

**Comment by Noordfrees** [comment]
During creating a normal game, the players are first created directly after loading the tribes, later destroyed when `Reloading the game from replay`, then recreated while `Reading Player Info`. In a scenario, they are initially created at the same time and then *not* destroyed/rebuilt. When enabling replays for scenarios the behaviour is the same as in normal games. In all cases the AI instances are only created when the game is completely ready. At this point the `Player::AiPersistentState` data is identical between normal games and scenarios (except for irrelevant differences in neuron values etc). It is saveloaded in the same way for normal games and scenarios with replay. Hm…

**Comment by matthiakl** [comment]
Maybe #4656 makes a difference? The scenario script is run after the AI finishes initialization and may register new units without notifying the AI. After save reloading those are already part of the map and known to the AI.

**Comment by Noordfrees** [comment]
I already tried delaying the creation of the AI for a few seconds or minutes to ensure there was time for all scripts to run, but it made no difference :(

**Comment by hessenfarmer** [comment]
well with intensive debug logging at last I found a major difference which points to an initialization problem:
- in a normal game the Ai is only evaluating its own military buildings for construction, while in a scenario all military buildings of each tribe are evaluated so if a building of a different tribe gets the highest score nothing is built. this is the reason why frisian scenario AI performs a bit better as mostly frisian buildings get the highes rate
- I believe the reason is that the second part of this line https://github.com/widelands/widelands/blob/fe7992d34d0336f88e1b1b33bc5ab2ec1d56839a/src/ai/ai_help_structs.cc#L412 is not correctly evaluated.
-  this line is affected by our famous allow_buildings(all) lines in the scripts where I belive the error comes from.   

@Noordfrees @matthiakl  please verify my assumption and help in elaborating a solution

**Comment by Noordfrees** [comment]
Looks like you found it 🚀 🚀 🚀 
If I add `&& p.tribe().has_building(id)` to this line then the AI players seem to behave normal in fri03 and fri04

**Comment by hessenfarmer** [comment]
I added the check for having tribes buildings at the place where it is necessary rather then in the definition of the buildable property to interfere as low as possible with other places in the code. If we change the property we need to analyze all other places where this is used. If I have time I'll try to analyze them whether the definition is wrong in all cases and the definition should be corrected.  
However I added this to my AI branch which makes #4995 even more urgent for me as after we have this in master we need to retest all scenarios again to evaluate their difficulty.
